api/src/main/java/io/opentelemetry/common/ImmutableKeyValuePairs.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2020, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.common;
+
+import static io.opentelemetry.internal.Utils.checkArgument;
+import static io.opentelemetry.internal.Utils.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * An immutable set of key-value pairs. Keys are only {@link String} typed. Can be iterated over
+ * using the {@link #forEach(KeyValueConsumer)} method.
+ *
+ * @param <V> The type of the values contained in this.
+ * @see Labels
+ * @see Attributes
+ */
+@Immutable
+abstract class ImmutableKeyValuePairs<V> {
+
+  List<Object> data() {
+    return Collections.emptyList();
+  }
+
+  /** Iterates over all the key-value pairs of attributes contained by this instance. */
+  @SuppressWarnings("unchecked")
+  public void forEach(KeyValueConsumer<V> consumer) {
+    for (int i = 0; i < data().size(); i += 2) {
+      consumer.consume((String) data().get(i), (V) data().get(i + 1));
+    }
+  }
+
+  static List<Object> sortAndFilter(Object[] data) {
+    checkArgument(
+        data.length % 2 == 0, "You must provide an even number of key/value pair arguments.");
+
+    quickSort(data, 0, data.length - 2);
+    return dedupe(data);
+  }
+
+  private static void quickSort(Object[] data, int leftIndex, int rightIndex) {
+    if (leftIndex >= rightIndex) {
+      return;
+    }
+
+    String pivotKey = (String) data[rightIndex];
+    int counter = leftIndex;
+
+    for (int i = leftIndex; i <= rightIndex; i += 2) {
+      if (((String) data[i]).compareTo(pivotKey) <= 0) {
+        swap(data, counter, i);
+        counter += 2;
+      }
+    }
+
+    quickSort(data, leftIndex, counter - 4);
+    quickSort(data, counter, rightIndex);
+  }
+
+  private static List<Object> dedupe(Object[] data) {
+    List<Object> result = new ArrayList<>(data.length);
+    Object previousKey = null;
+
+    for (int i = 0; i < data.length; i += 2) {
+      Object key = data[i];
+      Object value = data[i + 1];
+      checkNotNull(key, "You cannot provide null keys for creation of attributes.");
```

Comment:
hmm. null keys seems like a pretty bad problem. what do you suggest we do if some instrumentation puts in a null key?

Reply:
I agree it's a bad problem but eventually some user will find their app crash because of observability logic and complain even in the most unlikely scenarios :) And I think that's the motivation behind those principles.

Just skipping of course makes finding these bugs hard - how about a configurable check that can be enabled in unit tests? https://github.com/aws/aws-xray-sdk-java/blob/master/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/utils/LooseValidations.java#L24

Reply:
How about skipping and logging? I do think this will be a very rare occurrence, and will almost certainly indicate a programming bug.

Reply:
ok, skipped null keys and logged.